client/mm/mm_arb_market_maker.go

```diff
@@ -0,0 +1,651 @@
+	// cexTradeUpdatesID is passed to the Trade function of the cex
+	// so that the cex knows to send update notifications for the
+	// trade back to this bot.
+	cexTradeUpdatesID int
+	core              clientCore
+	log               dex.Logger
+	cfg               *ArbMarketMakerConfig
+	mkt               *core.Market
+	book              dexOrderBook
+	rebalanceRunning  atomic.Bool
+	currEpoch         atomic.Uint64
+
+	ordMtx         sync.RWMutex
+	ords           map[order.OrderID]*core.Order
+	oidToPlacement map[order.OrderID]int
+
+	matchesMtx  sync.RWMutex
+	matchesSeen map[order.MatchID]bool
+
+	cexTradesMtx sync.RWMutex
+	cexTrades    map[string]uint64
+
+	feesMtx  sync.RWMutex
+	buyFees  *orderFees
+	sellFees *orderFees
+}
+
+// groupedOrders returns the buy and sell orders grouped by placement index.
+func (m *arbMarketMaker) groupedOrders() (buys, sells map[int][]*groupedOrder) {
+	m.ordMtx.RLock()
+	defer m.ordMtx.RUnlock()
+	return groupOrders(m.ords, m.oidToPlacement, m.mkt.LotSize)
+}
+
+func (a *arbMarketMaker) handleCEXTradeUpdate(update *libxc.TradeUpdate) {
+	a.log.Debugf("CEX trade update: %+v", update)
+
+	if update.Complete {
+		a.cexTradesMtx.Lock()
+		delete(a.cexTrades, update.TradeID)
+		a.cexTradesMtx.Unlock()
+		return
+	}
+}
+
+// processDEXMatch checks to see if this is the first time the bot has seen
+// this match. If so, it sends a trade to the CEX.
+func (a *arbMarketMaker) processDEXMatch(o *core.Order, match *core.Match) {
+	var matchID order.MatchID
+	copy(matchID[:], match.MatchID)
+
+	a.matchesMtx.Lock()
+	if _, seen := a.matchesSeen[matchID]; seen {
+		a.matchesMtx.Unlock()
+		return
+	}
+	a.matchesSeen[matchID] = true
+	a.matchesMtx.Unlock()
+
+	var cexRate uint64
+	if o.Sell {
+		cexRate = uint64(float64(match.Rate) / (1 + a.cfg.Profit))
+	} else {
+		cexRate = uint64(float64(match.Rate) * (1 + a.cfg.Profit))
+	}
+	cexRate = steppedRate(cexRate, a.mkt.RateStep)
+
+	a.cexTradesMtx.Lock()
+	defer a.cexTradesMtx.Unlock()
+
+	tradeID, err := a.cex.Trade(a.ctx, dex.BipIDSymbol(a.base), dex.BipIDSymbol(a.quote), !o.Sell, cexRate, match.Qty, a.cexTradeUpdatesID)
+	if err != nil {
+		a.log.Errorf("Error sending trade to CEX: %v", err)
+		return
+	}
+
+	// Keep track of the epoch in which the trade was sent to the CEX. This way
+	// the bot can cancel the trade if it is not filled after a certain number
+	// of epochs.
+	a.cexTrades[tradeID] = a.currEpoch.Load()
+}
+
+func (a *arbMarketMaker) processDEXMatchNote(note *core.MatchNote) {
+	var oid order.OrderID
+	copy(oid[:], note.OrderID)
+
+	a.ordMtx.RLock()
+	o, found := a.ords[oid]
+	if !found {
+		a.ordMtx.RUnlock()
+		return
+	}
+	a.ordMtx.RUnlock()
+
+	a.processDEXMatch(o, note.Match)
+}
+
+func (a *arbMarketMaker) processDEXOrderNote(note *core.OrderNote) {
+	var oid order.OrderID
+	copy(oid[:], note.Order.ID)
+
+	a.ordMtx.Lock()
+	o, found := a.ords[oid]
+	if !found {
+		a.ordMtx.Unlock()
+		return
+	}
+	a.ords[oid] = note.Order
+	a.ordMtx.Unlock()
```

Comment:
```suggestion
	a.ordMtx.RLock()
	o, found := a.ords[oid]
	a.ordMtx.RUnlock()
	if !found {
		return
	}
	a.ords[oid] = note.Order
```

Reply:
But then `a.ords` won't be guarded when doing `a.ords[oid] = note.Order`.